trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/rest/TrinoS3ProxyResource.java

```diff
@@ -13,19 +13,80 @@
  */
 package io.trino.s3.proxy.server.rest;
 
+import com.google.common.base.Splitter;
+import com.google.inject.Inject;
+import io.trino.s3.proxy.server.credentials.Credentials;
+import io.trino.s3.proxy.server.credentials.SigningController;
+import io.trino.s3.proxy.server.credentials.SigningMetadata;
 import jakarta.ws.rs.GET;
+import jakarta.ws.rs.HEAD;
 import jakarta.ws.rs.Path;
-import jakarta.ws.rs.Produces;
-import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.PathParam;
+import jakarta.ws.rs.WebApplicationException;
+import jakarta.ws.rs.container.AsyncResponse;
+import jakarta.ws.rs.container.Suspended;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.Response;
+import org.glassfish.jersey.server.ContainerRequest;
 
-@Path(TrinoS3ProxyRestConstants.BASE_PATH)
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@Path(TrinoS3ProxyRestConstants.S3_PATH)
 public class TrinoS3ProxyResource
 {
+    private final SigningController signingController;
+    private final TrinoS3ProxyClient proxyClient;
+
+    @Inject
+    public TrinoS3ProxyResource(SigningController signingController, TrinoS3ProxyClient proxyClient)
+    {
+        this.signingController = requireNonNull(signingController, "signingController is null");
+        this.proxyClient = requireNonNull(proxyClient, "proxyClient is null");
+    }
+
     @GET
-    @Path("hello")
-    @Produces(MediaType.APPLICATION_JSON)
-    public String hello()
+    public void s3Get(@Context ContainerRequest request, @Suspended AsyncResponse asyncResponse)
+    {
+        s3Get(request, asyncResponse, "/");
+    }
+
+    @GET
+    @Path("{path:.*}")
+    public void s3Get(@Context ContainerRequest request, @Suspended AsyncResponse asyncResponse, @PathParam("path") String path)
+    {
+        proxyClient.proxyRequest(validateRequest(request), request, asyncResponse, getBucket(path));
+    }
+
+    @HEAD
+    public void s3Head(@Context ContainerRequest request, @Suspended AsyncResponse asyncResponse)
+    {
+        s3Head(request, asyncResponse, "/");
+    }
+
+    @HEAD
+    @Path("{path:.*}")
+    public void s3Head(@Context ContainerRequest request, @Suspended AsyncResponse asyncResponse, @PathParam("path") String path)
+    {
+        proxyClient.proxyRequest(validateRequest(request), request, asyncResponse, getBucket(path));
+    }
+
+    private String getBucket(String path)
+    {
+        List<String> parts = Splitter.on("/").splitToList(path);
+        return parts.isEmpty() ? "" : parts.getFirst();
```

Comment:
Currently we are prefixing all operations with `S3_PATH` and we expect the bucket to be the first part of the remainder of the path.

I take it this would not support [virtual-host style buckets](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) right?

Happy to add that separately, just checking my understanding of the code :)

Reply:
This is due to my ignorance of virtual-host buckets. Shall we address that now or in a future PR?

Reply:
Let's do in a future PR, opened #19